### PR TITLE
[AuK] Store the DiT in bf16 and build the attention bias once per step

### PR DESCRIPTION
**Pull request by Dayuxiaoshui**
## Motivation

The AuK base checkpoint runs 32 Euler steps with CFG, so each request costs 64 DiT evaluations of a 1.53B-parameter Flux-style backbone. On one H200 with SeedTTS-en voice cloning (about 4 s of audio per request) the initial implementation served a single request in 1.9 s and throughput barely moved with concurrency: 0.53 qps at c=1 and 1.31 qps at c=16, while latency grew six-fold.

`torch.profiler` on one CFG step at bs=8 (500 target + 250 reference frames) attributed the time as follows:

- The DiT was stored in fp32 and run under bf16 autocast. Autocast does not cache casts of `requires_grad=False` weights, so every `linear` re-read 6 GB of fp32 weights and wrote a bf16 copy: about 45 ms of `copy_` / `to` per step against about 31 ms of actual matmul.
- Every attention call expanded a boolean key mask to `[B, H, Q, K]`; SDPA then converted it to a float mask with `aten::where` (24 ms per step across the 60 calls) and could never select the flash kernel, since flash rejects any `attn_mask`.
- The q/k `RMSNorm` and the AdaLN `LayerNorm` ran in fp32 under autocast.

Both costs scale with batch size, which is why batching helped so little: a bs=8 step cost about 4x a bs=1 step at these lengths.

## Modifications

- `create_auk_engine_executor` gains a `weight_dtype` factory argument. `AuKPipelineConfig` sets `bfloat16`: the backbone is stored in bf16 and runs without autocast. Only `flow.transformer` is converted; the conditioning stage shares the same cached `AuKFlowMatching` object for `flow.fuse` and keeps those layer weights fp32. Leaving `weight_dtype` unset (or `float32`) keeps the fp32-weights + autocast recipe, which is what `tests/test_model/test_auk_parity.py` compares against upstream.
- `TimestepEmbedding` keeps the sinusoid in fp32 (its arguments reach about 1000 rad, which bf16 cannot represent) and casts only the MLP input.
- `AuKFlowMatching.sample_batch` casts the text conditioning, reference latent and current state to the backbone dtype once per call. The ODE state stays fp32 through type promotion (`y_fp32 + dt * v_bf16` is fp32), so no explicit cast is needed and the legacy path is unchanged.
- Key-padding masks become one additive `[B, 1, 1, K]` float bias per forward, built in `AuKDit.forward` and shared by the 10 double-stream and 20 single-stream blocks. The per-block `attn_mask_enabled` plumbing no longer had a reader and was removed; the flag stays on `AuKDit`.
- `docs/cookbook/auk.md` documents the two dtype modes; a unit test checks that a bf16 backbone returns fp32, finite latents that track the fp32 backbone.

## Related Issues

Follow-up to #2061 (initial AuK support). No issue is closed by this PR.

## Accuracy Test

- **Legacy path is bit-identical.** With `weight_dtype` unset, 32-step latents for four batched requests and one single request are `torch.equal` to those produced by the pre-change code on the same inputs and seeds. The attention bias rewrite therefore changes no upstream-comparable numerics.
- **Unit tests.** `pytest tests/unit_test/auk`: 36 passed, including the existing check that batched, padded sampling matches per-request sampling within 1e-5.
- **End-to-end WER**, 48 SeedTTS-en voice-clone samples, seed 1234, transcribed with `Qwen/Qwen3-ASR-1.7B` via `benchmark_tts_seedtts --transcribe-only`:

  | path | corpus WER | per-sample mean | samples with WER > 0 |
  |---|---|---|---|
  | fp32 weights + autocast | 0.55% | 0.56% | 2 |
  | bf16 weights | 0.55% | 0.56% | 2 (same samples, same errors) |

  All 48 transcripts are identical between the two paths.
- **UTMOS** (`--utmos-only`, same 48 clips): 3.815 (legacy) vs 3.798 (bf16).
- The waveforms themselves differ (mean SNR 14 dB, mean correlation 0.93): bf16 rounding is amplified over 32 Euler steps. The same order of deviation already exists between batched and single-request sampling on the legacy path, so this is not a new property.
- The upstream parity test was not run here; it needs the upstream checkout and is unaffected because it exercises the unchanged legacy path.

## Benchmark & Profiling

Hardware: one NVIDIA H200 (GPU 0 of a shared 8-GPU host), torch 2.13, commit `428b7322` as baseline. Dataset: `zhaochenyang20/seed-tts-eval-arrow`, EN, voice cloning with reference audio and transcript, seed 1234.

### Standalone DiT step (CFG, 2x batch inside), ms per step

Measured with `AuKFlowMatching.sample_batch` on an idle GPU 0; "legacy" is fp32 weights + bf16 autocast, "new" is bf16 weights without autocast.

| target + ref frames | bs | legacy | new | speedup |
|---|---|---|---|---|
| 150 + 150 | 8 | 64.0 | 55.2 | 1.16x |
| 150 + 150 | 16 | 115.8 | 85.1 | 1.36x |
| 300 + 200 | 8 | 94.9 | 69.4 | 1.37x |
| 300 + 200 | 16 | 176.7 | 132.0 | 1.34x |
| 500 + 500 | 8 | 174.4 | 131.8 | 1.32x |
| 500 + 500 | 16 | 332.3 | 251.9 | 1.32x |

Server-like conditions do not change the picture: with per-request lengths drawn fresh for every batch and the DiT on a non-default CUDA stream, bs=16 costs 396 ms (legacy) vs 292 ms (new) per step and bs=1 costs 39 vs 34 ms; there is no cold-shape penalty from the cuDNN SDPA backend. With a concurrent fp32 VAE-decode thread on another stream, both slow down by about 35% (528 vs 410 ms) and the ordering is unchanged.

### Server sweep

Absolute numbers on this host swing by up to 50% with other tenants' CPU load (the same legacy configuration measured 1.60 and 1.05 qps at c=32 half an hour apart), so the comparison below alternates the two configurations on the same GPU: legacy, new, legacy, new, each a fresh server, 64 samples per level.

| run | path | c=16 mean / p95 latency | c=16 qps | c=32 mean / p95 latency | c=32 qps |
|---|---|---|---|---|---|
| 1 | legacy | 12.97 / 15.31 s | 1.13 | 26.85 / 34.99 s | 1.05 |
| 2 | new | 10.37 / 13.52 s | 1.43 | 20.58 / 29.31 s | 1.35 |
| 3 | legacy | 13.41 / 17.18 s | 1.10 | 26.82 / 34.55 s | 1.05 |
| 4 | new | 11.49 / 16.58 s | 1.32 | 17.94 / 22.87 s | 1.54 |

At c=1 (96 samples, sequential runs) mean latency went from 1.70 s to 1.63 s and p95 from 1.89 s to 1.83 s; the DiT is kernel-launch bound at bs=1 (35-40 ms per step regardless of sequence length), so the weight-cast savings are small there. c=64 was measured only un-interleaved and is inconclusive under this host's noise.

### Where the remaining time goes

After the change, one bs=8 step (500 + 250 frames) spends about 40% of GPU time in unfused elementwise kernels and `LayerNorm`, about 20% in matmul and about 10% in attention. `aten::where` and the autocast `copy_` kernels are gone. The next levers are fusing rope / norms (torch.compile brought bs=1 from 33 to 25 ms per step in a trial), CUDA graphs with bucketed padding for the launch-bound bs=1 case, and length-grouped batching.

### Reproduce

```bash
# serve (default: bf16 weights)
python -m sglang_omni.cli serve --model-path tencent/AuK --port 8000
# legacy path for comparison
python -m sglang_omni.cli serve --model-path tencent/AuK --port 8000 \
  --auk_engine.factory.weight_dtype float32

# latency / throughput sweep against a running server
python -m benchmarks.eval.benchmark_tts_seedtts --model tencent/AuK --port 8000 \
  --use-existing-server --generate-only --concurrencies 1,4,8,16,32 \
  --max-samples 96 --warmup 4 --output-dir results/auk_sweep

# WER on the generated audio (starts a Qwen3-ASR server on --port)
python -m benchmarks.eval.benchmark_tts_seedtts --model tencent/AuK --port 8010 \
  --transcribe-only --lang en --output-dir results/auk_sweep/c1
```

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the `run-ci` label. The TTS CI selectors (`higgs`, `moss`, `qwen3-tts`) do not cover AuK; the AuK unit tests run in the default suite and the checkpoint parity test is opt-in through `AUK_UPSTREAM_SOURCE` / `AUK_PARITY_CHECKPOINT`.

### CODE REVIEW

File: sglang_omni/models/auk/stages.py [hunk 1]
```diff
@@ -216,11 +227,22 @@ def create_auk_engine_executor(
...
+    autocast (see docs/cookbook/auk.md, Sampling).
+    """
     checkpoint = resolve_checkpoint(model_path)
     config = make_runtime_config(checkpoint)
     device = torch.device(resolve_device_spec(device, gpu_id))
     flow = _load_flow(checkpoint, str(device))
+    if weight_dtype not in (None, "float32"):
+        # Backbone only: the conditioning stage shares this cached object for
+        # ``flow.fuse`` and keeps those layer weights fp32.
+        flow.transformer.to(dtype=_COMPUTE_DTYPES[weight_dtype])
```
Comment (AkazaAkane) [inline sglang_omni/models/auk/stages.py]: _load_flow is @lru_cached, so after one executor is created with weight_dtype="bfloat16", flow.transformer.to(...) permanently converts the shared cached transformer. A later executor for the same checkpoint/device with weight_dtype=None or "float32" will receive that BF16 transformer and _dit_compute_context() will also skip autocast because it checks the actual transformer dtype.

For example:

create_auk_engine_executor(..., weight_dtype="bfloat16")
create_auk_engine_executor(..., weight_dtype="float32")

does not restore the second executor to the documented FP32-weights + BF16-autocast path.

**Review by PoTaTo-Mika — APPROVED** [review]
illegal dtype will raise a KeyError, maybe apply more detailed notifications is better?
